granary/reddit.py

```diff
@@ -0,0 +1,215 @@
+"""reddit.com source class.
+"""
+
+from . import source
+import datetime
+import logging
+from oauth_dropins import reddit
+from oauth_dropins.webutil import util
+import re
+import urllib.parse, urllib.request
+
+import praw
+
+class Reddit(source.Source):
+
+    DOMAIN = 'reddit.com'
+    BASE_URL = 'https://reddit.com'
+    NAME = 'reddit'
+
+    def __init__(self, refresh_token):
+        self.refresh_token = refresh_token
+
+    @staticmethod
+    def timestamp_to_iso8601(time_stmp):
+        """Converts a timestamp from UNIX format to ISO 8601.
+        """
+        if not time_stmp:
+            return None
+
+        return datetime.datetime.fromtimestamp(time_stmp).isoformat()
+
+    def redditor_to_actor(self, user):
+        """Converts a praw Redditor to an actor.
+        https://praw.readthedocs.io/en/latest/code_overview/models/redditor.html
+
+        Args:
+          user: praw Redditor object
+
+        Returns:
+          an ActivityStreams actor dict, ready to be JSON-encoded
+        """
+        username = user.name
+        if not username:
+            return {}
+
+
+        # trying my best to grab all the urls from the profile description
+
+        description = ''
+        if user.subreddit:
+            user_url = self.BASE_URL+user.subreddit.get('url')
+            urls = [user_url]
+            description = user.subreddit['public_description']
+            profile_urls = re.findall(r'(https?://\S+)', description)
+            profile_urls += re.findall(r'(http?://\S+)', description)
+            urls += util.trim_nulls(profile_urls)
+        else:
+            urls = [self.BASE_URL+'/user/'+username]
+
+        image = user.icon_img
+
+        return util.trim_nulls({
+          'objectType': 'person',
+          'displayName': username,
+          'image': {'url': image},
+          'id': self.tag_uri(username),
+          # numeric_id is our own custom field that always has the source's numeric
+          # user id, if available.
+          'numeric_id': user.id,
+          'published': self.timestamp_to_iso8601(user.created_utc),
+          'url': urls[0],
+          'urls': [{'value': u} for u in urls] if len(urls) > 1 else None,
+          'username': username,
+          'description': description,
+          })
+
+
+    def praw_to_object(self, thng, typ):
+        """
+        Converts a tweet to an object.
+
+        Args:
+          thng: a praw object, Sumbission or Comment
+          typ: string to denote whether to get submission or comment content
+
+        Returns:
+          an ActivityStreams object dict, ready to be JSON-encoded
+            """
+        obj = {}
+
+        # always prefer id_str over id to avoid any chance of integer overflow.
+        # usually shouldn't matter in Python, but still.
+        id = thng.id
+        if not id:
+            return {}
+
+        created_at = thng.created_utc
+        try:
+            published = self.timestamp_to_iso8601(created_at)
```

Comment:
oh also conveniently `webutil.util.maybe_timestamp_to_iso8601()` returns input unchanged if it fails, so you'll be able to drop this try/except.